### ISSUE TITLE
envClass의 env.py에 대한 질문

### ISSUE THREAD

**Issue by ByungSunBae** [kooock/pikaRL#7]
안녕하세요 현재 env.py 코드를 보고있습니다.

computeReward 함수 부분에서

1) 어느 한쪽이 15점일때 굳이 100점을 주지 않아도 될거같습니다. Pong-v0의 경우가 피카츄 배구처럼 컴퓨터 vs Agent인데 그쪽 환경에서도 100점으로 크게 주지 않았습니다. 그래도 학습은 잘됩니다.

즉, 
 

```
#game finish
if self.comScoreBuffer == 15:
       reward3 = -100
       self.isFinished = True

if self.agentScoreBuffer == 15:
        reward3 = 100
        self.isFinished = True
```
이 부분을

```
if self.comScoreBuffer == 15 or self.agentScoreBuffer == 15:
        self.isFinished = True
```
로만 바꾸시면 될듯하네요. 안바꾸셔도 되긴하지만, reward간의 격차가 크면 Q value가 크게 왔다갔다 할수도 있습니다.

2) 연속된 4개의 이미지를 사용하지 않았군요 (이게 가장 큰 문제일수도 있습니다.)
 state는 4개의 연속된 이미지를 사용해서 만들어야합니다.
그래야 움직이는 것을 제대로 학습합니다. channel이 3이어도 이건 똑같은 이미지의 Red, Green, Blue라서 가만히 정지된 사진만 보고 움직임을 판단하는 겁니다.

128X128X3의 이미지를 resizing과 gray scaling을 거쳐서 80X80X1의 이미지를 4개로 겹치세요
즉, 

```
첫번째 128X128X3     =====>     80X80X1---------|
두번째 128X128X3     =====>     80X80X1---------|_____> 80X80X4 (이게 state 1개입니다.)
세번째 128X128X3     =====>     80X80X1---------|
네번째 128X128X3     =====>     80X80X1---------|
```

모델에 들어가는 state가 1X80X80X4 (개수 X 높이 X 너비 X 채널 /  여기서 채널은 RGB가 아닌 연속된 4개의 이미지를 뜻함)이므로 network의 구조도 Channel수를 3이 아니라 4로 셋팅해야합니다.

이렇게 구성된 state를 s라 하면,
state transition은 [s, a, r, s'] 이렇게 1개가 되어서 replay buffer로 들어가야합니다.

replay buffer 내부는 아래와 같습니다.
[ [s1, a1, r1, s2], [s2, a2, r2, s3], [s3, a3, r3, s4], .... ] (총 5만개가 있다고 가정)
여기서 배치사이즈만큼 랜덤 샘플링합니다.
배치사이즈는 5개라고 하겠습니다.

```
[s3, a3, r3, s4] (r3 = 0)
[s10, a10, r10, s11] (r10 = 1)
[s23, a23, r23, s24] (r23 = 0)
[s18, a18, r18, s19] (r18 = 0)
[s7, a7, r7, s8] (r7 = -1)
```
위의 데이터를 가지고 model을 업데이트 해야하는데, 이때 reward가 -1 혹은 1점이면
에피소드가 끝나지 않았더라도 target value를 계산할때는 구분을 지어야 하기 때문에 terminal로 처리합니다. 참고로 현재 이 글에서 target value를 계산하는 방식은 이 피카츄배구환경에 맞춰서 설명을 드린것이기 때문에 다른 환경에선 다를 수 있습니다. 즉 게임의 구분을 어떻게 짓느냐에 따라 terminal로 처리하는 것이 달라집니다. 물론 에피소드가 끝나는때는 terminal로 처리해야하구요.
즉, 위의 배치사이즈의 데이터를 가지고 target value를 계산한다면 아래와 같습니다.

(Q'는 target network로 a'은 몰라도 됩니다. 어차피 모델이 다음 state에 대해서 각 action별로 Q value를 추정하고 그값들중 가장 큰값을 사용하기 때문입니다.)

```
[s3, a3, r3, s4] (r3 = 0)  ==> target_value = r3 + maxQ'(s4, a')
[s10, a10, r10, s11] (r10 = 1)  ==> target_value = r10 (s11은 종료된걸로 간주)
[s23, a23, r23, s24] (r23 = 0)  ==> target_value = r23 +  maxQ'(s24, a')
[s18일때도 s3와 s23일때 처럼 target_value 계산]
[s7, a7, r7, s8] (r7 = -1) ==> target_value = r7 (s8은 종료된걸로 간주)
```

target network의 업데이트 주기는 10이 아니라 100 혹은 150으로 설정해보세요.
target network가 자주 업데이트되면 마치 움직이는 과녁을 맞추는 것과 비슷하니까요

이 이후부터는 원래 하시던대로 하면 됩니다.

**Comment by kooock** [comment]
와 정말 큰 도움이 되었어요 사실 논문을 읽어보지 않고 김성훈 교수님 강의만 보고 한건데 많이 놓친부분이 많네요 논문 제대로 분석해야겠네요

**Comment by ByungSunBae** [comment]
그리고 또 하나 있습니다. 이거는 페북에 이웅원님이 달아놓은 댓글보고 말씀드리는건데 target value가 real value이고 Q value 자체를 추정하는 것이기 때문에 Loss function은 MSE로 하셔야합니다. Cross Entropy는 classification할때 확률값과 정답label간의 차이를 줄여주는 것입니다. MSE로 꼭 수정하시기 바랍니다!

**Comment by kooock** [comment]
4프레임은 어떤 기준으로 잡아야 할까요 액션 후 4프레임 인가요?
